pkg/reconciliation/reconcile_racks.go

```diff
@@ -2092,16 +2094,152 @@ func (rc *ReconciliationContext) CheckConditionInitializedAndReady() result.Reco
+	})
+
+	for idx, pod := range rc.dcPods {
+		features, err := rc.NodeMgmtClient.FeatureSet(pod)
+		if err != nil {
+			return result.Error(err)
+		}
+
+		if pod.Annotations == nil {
+			pod.Annotations = make(map[string]string)
+		}
 
-	for idx := range rc.dcPods {
-		err = rc.NodeMgmtClient.CallKeyspaceCleanupEndpoint(rc.dcPods[idx], -1, "", nil)
-		if err == nil {
-			break
+		if podJobId, found := pod.Annotations[podJobIdAnnotation]; found {
+			if features.Supports(httphelper.AsyncSSTableTasks) {
+				_, found := pod.Annotations[podJobStatusAnnotation]
+				if !found {
+					// Pod is currently processing something, or has finished processing.. update the status
+					details, err := rc.NodeMgmtClient.JobDetails(pod, podJobId)
+					if err != nil {
+						rc.ReqLogger.Error(err, "Could not get JobDetails for pod", "Pod", pod)
+						return result.Error(err)
+					}
+
+					if details.Id == "" {
+						// This job was not found, pod most likely restarted. Let's retry..
+						delete(pod.Annotations, podJobIdAnnotation)
+						err = rc.Client.Update(context.Background(), pod)
+						if err != nil {
+							return result.Error(err)
+						}
+						return result.RequeueSoon(2)
+					} else if details.Status == podJobError {
+						// Log the error, move on
+						rc.ReqLogger.Error(fmt.Errorf("cleanup failed: %s", details.Error), "Pod", pod)
+						pod.Annotations[podJobStatusAnnotation] = podJobError
+						err = rc.Client.Update(context.Background(), pod)
+						if err != nil {
+							return result.Error(err)
+						}
+						continue
+					} else if details.Status == podJobCompleted {
+						// Pod has finished, remove the job_id and let us move to the next pod
+						pod.Annotations[podJobStatusAnnotation] = podJobCompleted
+						err = rc.Client.Update(context.Background(), pod)
+						if err != nil {
+							return result.Error(err)
+						}
+						continue
+					} else if details.Status == podJobWaiting {
+						// Job is still running or waiting
+						return result.RequeueSoon(2)
+					}
+				} else {
+					// This pod has finished since it has a status
+					continue
+				}
+			} else {
+				if len(jobRunner) > 0 {
+					// Something is still holding the worker
+					return result.RequeueSoon(2)
+				}
+
+				// Nothing is holding the job, has this pod finished or hasn't it ran anything?
+				if _, found := pod.Annotations[podJobStatusAnnotation]; found {
+					// Pod has finished, check next one
+					continue
+				}
+			}
 		}
+
+		if features.Supports(httphelper.AsyncSSTableTasks) {
+			// Pod isn't running anything at the moment, this pod should run next
+			jobId, err := rc.NodeMgmtClient.CallKeyspaceCleanup(pod, -1, "", nil)
+			if err != nil {
+				// We can retry this later, it will only restart the cleanup but won't otherwise hurt
+				return result.Error(err)
+			}
+			pod.Annotations[podJobHandler] = jobHandlerMgmtApi
+			pod.Annotations[podJobIdAnnotation] = jobId
+
+			err = rc.Client.Update(context.Background(), pod)
```

Comment:
Would `Patch` be better here? I don't think we should care about conflicts here. If the update fails due to a conflict and we requeue, do we wind up calling `rc.NodeMgmtClient.CallKeyspaceCleanup` again? And shouldn't be using `context.Background()`.

This makes me wonder if we should be setting a status right before calling `CallKeyspaceCleanup` to indicate that we are about to start the operation. We would check for that status before making the call. If it is set, then we should check to see if the job is already in progress before calling `CallKeyspaceCleanup`.

Reply:
We're checking the status before we start the job. The previous part:

``if podJobId, found := pod.Annotations[podJobIdAnnotation]; found {``

is ran before this, so if the podJobId is set, then we will not run this part of the code. We set the status here to indicate the jobId (so we can search for it), and we won't get that id before we actually send the request for a job.

If this Update fails and we come back to this part, then the new CallKeyspaceCleanup shouldn't really matter.